### ISSUE TITLE
Extra screen flash coming out of screensaver when "Keep screensaver on screen until tap" is on

### ISSUE THREAD

**Issue by melyux** [koreader/koreader#9285]
* KOReader version: 2022.6
* Device: Kindle PW5

#### Issue
I noticed when I enabled this feature, the screen flashes twice when coming out of the screensaver, the first flash happening when waking but keeping the screen the same (which needn't happen) until a tap.

I tried to trace the logic and was led here: https://github.com/koreader/koreader/blob/4f7ed7e7063625e441270d571cfd8cb3327b145a/frontend/device/kindle/device.lua#L248

Kindle:outofScreenSaver() sets the whole screen "dirty" and then later when we tap the screen to actually dismiss the screensaver, ScreenSaverWidget.onTap() eventually leads to setting the screen "dirty" again, which accounts for the second flash when actually dismissing the screensaver.

The first flash seems unnecessary since the screen content doesn't actually change until we tap and dismiss. This change seems to have happened with #6240 with the comment "redraw everything in case the framework managed to screw us over..."

Now I don't want the framework screwing us over, but the double full flash seems unnecessary in this case. What exactly does this prevent on Kindle? Would it be so terrible in the "keep screensaver on screen" mode for it to happen to save us a flash?

I was going to test on my device without that extra setDirty() line but haven't had the chance yet. Still gathering info. Thinking of also doing a PR with an indicator icon to differentiate between the true sleep and "awake but screensaver still on" states, maybe. Comments welcome though

#### Steps to reproduce
Have a Kindle. Turn on "Keep screensaver on screen" mode with "Dismiss with a tap", in Gear Settings > Screen > Screensaver. Sleep the Kindle. Unsleep the Kindle. Observe the Kindle screen flashing upon wake despite the screen content being exactly the same.

**Comment by NiLuJe** [comment]
IIRC, it's to workaround issues with awesome racing against us.

It's extremely firmware/device/environment-specific to a large extent, so I'm wary of touching that without extensive testing, especially when the drawback is fairly harmless ;).

**Comment by melyux** [comment]
I could do testing, but wary of "bricking" the Kindle and/or being forced into a factory reset. What happens if Awesome wins the race? (In fact, I don't really know what Awesome is, couldn't find much documentation of it outside of occasional code comments, but I assumed it was the Kindle OS).

**Comment by NiLuJe** [comment]
`awesome` is the WM (window manager) used by the stock OS. It happens to be heavily customizable because Lua, which is probably what appealed to lua126, as it made it relatively easy to chuck a whole lot of custom trickery in there, in particular handling all the various eInk refresh tricks.

TL;DR: It's what refreshes the screen for (almost) everything in the stock OS (the rest is left to `blanket`, a custom module).

So, what this attempts to do is preventing the stock OS from seeping through during a wake-up, in the form of a small/large/broken part of the stock OS being shown on screen.

**Comment by melyux** [comment]
I've been testing for the last month with the Kindle PW5 with the lines I mentioned removed and never had Awesome break through yet. (I have seen it break through in other places in KOReader, but never here.) Just a data point, but it could be good to remove these lines and see if anyone actually has problems without the double flash, since this precaution seems to hav been added preemptively?

https://github.com/koreader/koreader/blob/55c0fc6190d1be330b4efb9a8447982b6c52daee/frontend/device/kindle/device.lua#L256-L258

**Comment by NiLuJe** [comment]
Nope, it wasn't preemptive ;).

Like I said, ti's *extremely* device and firmware specific. While it might not be necessary on your device, it definitely was on others ;).

**Comment by NiLuJe** [comment]
That said, I'll double-check on a few devices, because I don't quite recall if I tested that with "keep ss on until tap" (or if that option even existed at the time) on said offending devices.

(i.e., it might make sense to skip that when that option is enabled).

**Comment by melyux** [comment]
Same goes for the new "wait until gesture" option as well. But when using the other options, no flash is needed either since there's already a flash later when dismissing the screensaver

**Comment by NiLuJe** [comment]
I'm actually concerned about what happens *before* the tap/gesture.

IIRC, the issue is the framework winning the race and drawing part/all of the stock screensaver ;).
